app/reaction.go

```diff
@@ -40,6 +40,12 @@ func (a *App) SaveReactionForPost(c *request.Context, reaction *model.Reaction)
 		}
 	}
 
+	if post.RootId == "" {
```

Comment:
There are couple of places, where this condition is not needed. It's the caller's business/responsibility to decide whether or not to call this function. 
`DeletePersistentNotificationsPost` handles one job of deleting as long it's allowed by the system/configs.

Reply:
Hmm, I couldn't find a case when `DeletePersistentNotificationsPost` is called without first checking the `post.RootId`. It /feels/ like it's part of the core business logic, and my primary concern is that it might get called without checking this bit if indeed that's currently core logic.

Reply:
* `Ack` doesn't need to check for rootPost, although it still can, but it's not needed as the `Ack` itself is not allowed in reply-posts. https://github.com/mattermost/mattermost-server/blob/fa472d439a262ea232f0241933f238a2aec89b37/app/post_acknowledgements.go#L44
* Condition itself may also differ as in this case, where I need to check that it must be a reply-post and the post arg passed won't be enough to validate that inside the `DeletePersistentNotificationsPost` function. https://github.com/mattermost/mattermost-server/blob/fa472d439a262ea232f0241933f238a2aec89b37/app/post.go#L388